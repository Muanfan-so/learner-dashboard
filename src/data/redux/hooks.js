import { useSelector } from 'react-redux';

import { actions as appActions } from './app/reducer';
import appSelectors from './app/selectors';
import requestSelectors from './requests/selectors';

const { courseCard } = appSelectors;

export const usePageNumber = () => useSelector(appSelectors.pageNumber);
export const useEmailConfirmationData = () => useSelector(appSelectors.emailConfirmation);
export const useEnterpriseDashboardData = () => useSelector(appSelectors.enterpriseDashboard);
export const usePlatformSettingsData = () => useSelector(appSelectors.platformSettings);
// suggested courses is max at 3 at the moment.
export const useSuggestedCoursesData = () => useSelector(appSelectors.suggestedCourses).slice(0, 3);
export const useRecommendedCoursesData = () => useSelector(appSelectors.recommendedCourses);
export const useSelectSessionModalData = () => useSelector(appSelectors.selectSessionModal);

export const useHasCourses = () => useSelector(appSelectors.hasCourses);
export const useHasAvailableDashboards = () => useSelector(appSelectors.hasAvailableDashboards);
export const useCurrentCourseList = (opts) => useSelector(
  state => appSelectors.currentList(state, opts),
);
export const useShowSelectSessionModal = () => useSelector(appSelectors.showSelectSessionModal);

// eslint-disable-next-line
export const useCourseCardData = (selector) => (cardId) => useSelector(
  (state) => selector(state, cardId),
);

export const useCardCertificateData = useCourseCardData(courseCard.certificate);
export const useCardCourseData = useCourseCardData(courseCard.course);
export const useCardCourseRunData = useCourseCardData(courseCard.courseRun);
export const useCardEnrollmentData = useCourseCardData(courseCard.enrollment);
export const useCardEntitlementData = useCourseCardData(courseCard.entitlement);
export const useCardGradeData = useCourseCardData(courseCard.gradeData);
export const useCardProviderData = useCourseCardData(courseCard.courseProvider);
export const useCardRelatedProgramsData = useCourseCardData(courseCard.relatedPrograms);

export const useUpdateSelectSessionModalCallback = (dispatch, cardId) => () => dispatch(
  appActions.updateSelectSessionModal(cardId),
);

export const useMasqueradeData = () => useSelector(requestSelectors.masquerade);

export const useIsPendingRequest = (requestName) => useSelector(requestSelectors.isPending(requestName));
