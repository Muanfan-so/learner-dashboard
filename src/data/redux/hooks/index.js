import * as app from './app';
import * as requests from './requests';

export * from './app';
export { requests };

export default { app, requests };
