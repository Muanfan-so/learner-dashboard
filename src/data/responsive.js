import { useWindowSize, breakpoints } from '@edx/paragon';
import { StrictDict } from 'utils';

const createResponsive = (fn) => () => {
  const { width } = useWindowSize();
  return fn(width);
};

export const isDesktopSize = createResponsive(
  (width) => width >= breakpoints.large.minWidth,
);

// Beware of using isPadSize() for styling. When the screen resizes pass the
// breakpoint, the screen will flash between the two styles. This is why it is
// recommened isDesktopSize() or isMobileSize() be used instead.
export const isIpadSize = createResponsive(
  (width) => width >= breakpoints.medium.minWidth && width < breakpoints.large.minWidth,
);

export const isMobileSize = createResponsive(
  (width) => width < breakpoints.medium.minWidth,
);

export default StrictDict({
  isDesktopSize,
  isIpadSize,
  isMobileSize,
});
