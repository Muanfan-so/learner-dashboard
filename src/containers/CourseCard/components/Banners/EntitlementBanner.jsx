import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from '@edx/frontend-platform/i18n';
import { Button, MailtoLink } from '@edx/paragon';

import { hooks as appHooks } from 'data/redux';
import { dateFormatter } from 'utils';

import Banner from 'components/Banner';
import useSelectSession from 'containers/SelectSession/hooks';
import messages from './messages';

export const EntitlementBanner = ({ courseNumber }) => {
  const {
    isEntitlement,
    hasSessions,
    isFulfilled,
    changeDeadline,
    showExpirationWarning,
    isExpired,
  } = appHooks.useCardEntitlementsData(courseNumber);
  const { supportEmail } = appHooks.usePlatformSettingsData();
  const { openSessionModal } = useSelectSession({ courseNumber });
  const { formatDate, formatMessage } = useIntl();

  if (!isEntitlement) {
    return null;
  }

  if (!hasSessions && !isFulfilled) {
    return (
      <Banner variant="warning">
        {formatMessage(messages.entitlementsUnavailable, {
          emailLink: supportEmail && <MailtoLink to={supportEmail}>{supportEmail}</MailtoLink>,
        })}
      </Banner>
    );
  }
  if (showExpirationWarning) {
    return (
      <Banner>
        {formatMessage(messages.entitlementsExpiringSoon, {
          changeDeadline: dateFormatter(formatDate, changeDeadline),
          selectSessionButton: (
            <Button variant="link" size="inline" className="m-0 p-0" onClick={openSessionModal}>
              {formatMessage(messages.selectSession)}
            </Button>
          ),
        })}
      </Banner>
    );
  }
  if (isExpired) {
    return (
      <Banner>
        {formatMessage(messages.entitlementsExpired)}
      </Banner>
    );
  }
  return null;
};
EntitlementBanner.propTypes = {
  courseNumber: PropTypes.string.isRequired,
};

export default EntitlementBanner;
