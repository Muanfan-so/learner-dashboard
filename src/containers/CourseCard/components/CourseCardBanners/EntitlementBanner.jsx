import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Button, MailtoLink } from '@edx/paragon';

import { hooks as appHooks } from 'data/redux';
import { useFormatDate } from 'utils/hooks';

import Banner from 'components/Banner';
import messages from './messages';

export const EntitlementBanner = ({ cardId }) => {
  const {
    isEntitlement,
    hasSessions,
    isFulfilled,
    changeDeadline,
    showExpirationWarning,
    isExpired,
  } = appHooks.useCardEntitlementData(cardId);
  const { supportEmail } = appHooks.usePlatformSettingsData();
  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(cardId);
  const { formatMessage } = useIntl();
  const formatDate = useFormatDate();

  if (!isEntitlement) {
    return null;
  }

  if (!hasSessions && !isFulfilled) {
    return (
      <Banner variant="warning">
        {formatMessage(messages.entitlementUnavailable, {
          emailLink: supportEmail && <MailtoLink to={supportEmail}>{supportEmail}</MailtoLink>,
        })}
      </Banner>
    );
  }
  if (showExpirationWarning) {
    return (
      <Banner>
        {formatMessage(messages.entitlementExpiringSoon, {
          changeDeadline: formatDate(changeDeadline),
          selectSessionButton: (
            <Button variant="link" size="inline" className="m-0 p-0" onClick={openSessionModal}>
              {formatMessage(messages.selectSession)}
            </Button>
          ),
        })}
      </Banner>
    );
  }
  if (isExpired) {
    return (
      <Banner>
        {formatMessage(messages.entitlementExpired)}
      </Banner>
    );
  }
  return null;
};
EntitlementBanner.propTypes = {
  cardId: PropTypes.string.isRequired,
};

export default EntitlementBanner;
