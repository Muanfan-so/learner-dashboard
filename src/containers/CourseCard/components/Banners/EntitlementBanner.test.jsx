import React from 'react';
import { shallow } from 'enzyme';

import { hooks as appHooks } from 'data/redux';
import EntitlementBanner from './EntitlementBanner';

jest.mock('components/Banner', () => 'Banner');
jest.mock('data/redux', () => ({
  hooks: {
    usePlatformSettingsData: jest.fn(),
    useCardEntitlementsData: jest.fn(),
  },
}));
jest.mock('containers/SelectSession/hooks', () => () => ({
  openSessionModal: jest.fn().mockName('useSelectSession.openSessionModal'),
}));

const courseNumber = 'my-test-course-number';

let el;

const entitlementsData = {
  isEntitlement: true,
  hasSessions: true,
  isFulfilled: false,
  changeDeadline: '11/11/2022',
  showExpirationWarning: false,
};
const platformData = { supportEmail: 'test-support-email' };

const render = (overrides = {}) => {
  const { entitlements = {} } = overrides;
  appHooks.useCardEntitlementsData.mockReturnValueOnce({ ...entitlementsData, ...entitlements });
  appHooks.usePlatformSettingsData.mockReturnValueOnce(platformData);
  el = shallow(<EntitlementBanner courseNumber={courseNumber} />);
};

describe('EntitlementBanner', () => {
  it('initializes data with course number from entitlements', () => {
    render();
    expect(appHooks.useCardEntitlementsData).toHaveBeenCalledWith(courseNumber);
  });
  test('no display if not an entitlement', () => {
    render({ entitlements: { isEntitlement: false } });
    expect(el.isEmptyRender()).toEqual(true);
  });
  test('snapshot: no sessions available', () => {
    render({ entitlements: { isFulfilled: false, hasSessions: false } });
    expect(el).toMatchSnapshot();
  });
  test('snapshot: expiration warning', () => {
    render({ entitlements: { showExpirationWarning: true } });
    expect(el).toMatchSnapshot();
  });
  test('no display if sessions available and not displaying warning', () => {
    render();
    expect(el.isEmptyRender()).toEqual(true);
  });
});
