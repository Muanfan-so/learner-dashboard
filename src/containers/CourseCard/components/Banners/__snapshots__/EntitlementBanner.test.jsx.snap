// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntitlementBanner snapshot: expiration warning 1`] = `
<Banner>
  <formatMessageFunction
    message={
      Object {
        "defaultMessage": "You must {selectSessionButton} by {changeDeadline} to access the course.",
        "description": "Entitlements course message when the entitlement is expiring soon.",
        "id": "learner-dash.courseCard.banners.entitlementsExpiringSoon",
      }
    }
    values={
      Object {
        "changeDeadline": "11/11/2022",
        "selectSessionButton": <Button
          className="m-0 p-0"
          onClick={[MockFunction useSelectSession.openSessionModal]}
          size="inline"
          variant="link"
        >
          select a session
        </Button>,
      }
    }
  />
</Banner>
`;

exports[`EntitlementBanner snapshot: no sessions available 1`] = `
<Banner
  variant="warning"
>
  <formatMessageFunction
    message={
      Object {
        "defaultMessage": "There are no sessions available at the moment.  The course team will create new sessions soon.  If no sessions appear, please contact {emailLink} for information.",
        "description": "Entitlements course message when no sessions are available",
        "id": "learner-dash.courseCard.banners.entitlementsUnavailable",
      }
    }
    values={
      Object {
        "emailLink": <MailtoLink
          to="test-support-email"
        >
          test-support-email
        </MailtoLink>,
      }
    }
  />
</Banner>
`;
